arwen/elrondapi/bigFloatOps.go

```diff
@@ -0,0 +1,650 @@
+	if arwen.WithFault(err, context, runtime.BigFloatAPIErrorShouldFailExecution()) {
+		return
+	}
+	managedType.ConsumeGasForBigFloatCopy(op)
+	op.Int(bigIntOp)
+	managedType.ConsumeGasForBigIntCopy(bigIntOp)
+	if op.IsInt() {
+		return
+	}
+	if bigIntOp.Sign() > 0 {
+		bigIntOp.Add(bigIntOp, big.NewInt(1))
+	}
+}
+
+//export v1_4_bigFloatTruncate
+func v1_4_bigFloatTruncate(context unsafe.Pointer, opHandle, bigIntHandle int32) {
+	managedType := arwen.GetManagedTypesContext(context)
+	metering := arwen.GetMeteringContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+
+	gasToUse := metering.GasSchedule().BigFloatAPICost.BigFloatTruncate
+	metering.UseGas(gasToUse)
+
+	op, err := managedType.GetBigFloat(opHandle)
+	if arwen.WithFault(err, context, runtime.BigFloatAPIErrorShouldFailExecution()) {
+		return
+	}
+	bigIntValue := managedType.GetBigIntOrCreate(bigIntHandle)
+
+	managedType.ConsumeGasForBigFloatCopy(op)
+	op.Int(bigIntValue)
+}
+
+//export v1_4_bigFloatSetInt64
+func v1_4_bigFloatSetInt64(context unsafe.Pointer, destinationHandle int32, value int64) {
+	managedType := arwen.GetManagedTypesContext(context)
+	metering := arwen.GetMeteringContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+
+	gasToUse := metering.GasSchedule().BigFloatAPICost.BigFloatSetInt64
+	metering.UseGas(gasToUse)
+
+	dest, err := managedType.GetBigFloatOrCreate(destinationHandle)
+	if arwen.WithFault(err, context, runtime.BigFloatAPIErrorShouldFailExecution()) {
+		return
+	}
+	dest.SetInt64(value)
+}
+
+//export v1_4_bigFloatIsInt
+func v1_4_bigFloatIsInt(context unsafe.Pointer, opHandle int32) int32 {
+	managedType := arwen.GetManagedTypesContext(context)
+	metering := arwen.GetMeteringContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+
+	gasToUse := metering.GasSchedule().BigFloatAPICost.BigFloatIsInt
+	metering.UseGas(gasToUse)
+	op, err := managedType.GetBigFloat(opHandle)
+	if arwen.WithFault(err, context, runtime.BigFloatAPIErrorShouldFailExecution()) {
+		return -1
+	}
+	managedType.ConsumeGasForBigFloatCopy(op)
+	if op.IsInt() {
+		return 1
+	}
+	return 0
+}
+
+//export v1_4_bigFloatSetBigInt
+func v1_4_bigFloatSetBigInt(context unsafe.Pointer, destinationHandle, bigIntHandle int32) {
+	managedType := arwen.GetManagedTypesContext(context)
+	metering := arwen.GetMeteringContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+
+	gasToUse := metering.GasSchedule().BigFloatAPICost.BigFloatSetBigInt
+	metering.UseGas(gasToUse)
+
+	dest, err1 := managedType.GetBigFloatOrCreate(destinationHandle)
+	bigIntValue, err2 := managedType.GetBigInt(bigIntHandle)
+	if arwen.WithMultipleFaults([]error{err1, err2}, context, runtime.BigFloatAPIErrorShouldFailExecution()) {
+		return
+	}
+	managedType.ConsumeGasForBigFloatCopy(dest)
+	managedType.ConsumeGasForBigIntCopy(bigIntValue)
+	dest.SetInt(bigIntValue)
+}
+
+//export v1_4_bigFloatMod
+func v1_4_bigFloatMod(context unsafe.Pointer, destinationHandle, op1Handle, op2Handle int32) {
```

Comment:
Floating point modulo is not yet as well-defined in programming as other operations... Hopefully this won't be an issue. To be on the safe side, I would suggest disabling this API function until we clarify it.

https://en.wikipedia.org/wiki/Modulo_operation

Reply:
https://pkg.go.dev/math#Mod